txnsync/incoming.go

```diff
@@ -0,0 +1,307 @@
+
+// makeIncomingMessageQueue creates an incomingMessageQueue object and initializes all the internal variables.
+func makeIncomingMessageQueue() incomingMessageQueue {
+	return incomingMessageQueue{
+		incomingMessages: make(chan incomingMessage, maxPeersCount),
+		enqueuedPeers:    make(map[*Peer]struct{}, maxPeersCount),
+	}
+}
+
+// getIncomingMessageChannel returns the incoming messages channel, which would contain entries once
+// we have one ( or more ) pending incoming messages.
+func (imq *incomingMessageQueue) getIncomingMessageChannel() <-chan incomingMessage {
+	return imq.incomingMessages
+}
+
+// enqueue places the given message on the queue, if and only if it's associated peer doesn't
+// appear on the incoming message queue already. In the case there is no peer, the message
+// would be placed on the queue as is.
+func (imq *incomingMessageQueue) enqueue(m incomingMessage) bool {
+	if m.peer != nil {
+		imq.enqueuedPeersMu.Lock()
+		defer imq.enqueuedPeersMu.Unlock()
+		if _, has := imq.enqueuedPeers[m.peer]; has {
+			return true
+		}
+	}
+	select {
+	case imq.incomingMessages <- m:
+		// if we successfully enqueued the message, set the enqueuedPeers so that we won't enqueue the same peer twice.
+		if m.peer != nil {
+			// at this time, the enqueuedPeersMu is still under lock ( due to the above defer ), so we can access
+			// the enqueuedPeers here.
+			imq.enqueuedPeers[m.peer] = struct{}{}
+		}
+		return true
+	default:
+		return false
+	}
+}
+
+// clear removes the peer that is associated with the message ( if any ) from
+// the enqueuedPeers map, allowing future messages from this peer to be placed on the
+// incoming message queue.
+func (imq *incomingMessageQueue) clear(m incomingMessage) {
+	if m.peer != nil {
+		imq.enqueuedPeersMu.Lock()
+		defer imq.enqueuedPeersMu.Unlock()
+		delete(imq.enqueuedPeers, m.peer)
+	}
+}
+
+// incomingMessageHandler
+// note - this message is called by the network go-routine dispatch pool, and is not synchronized with the rest of the transaction synchronizer
+func (s *syncState) asyncIncomingMessageHandler(networkPeer interface{}, peer *Peer, message []byte, sequenceNumber uint64) (err error) {
+	// increase number of incoming messages metric.
+	txsyncIncomingMessagesTotal.Inc(nil)
+
+	// check the return value when we exit this function. if we fail, we increase the metric.
+	defer func() {
+		if err != nil {
+			// increase number of unprocessed incoming messages metric.
+			txsyncUnprocessedIncomingMessagesTotal.Inc(nil)
+		}
+	}()
+
+	incomingMessage := incomingMessage{networkPeer: networkPeer, sequenceNumber: sequenceNumber, encodedSize: len(message), peer: peer}
+	_, err = incomingMessage.message.UnmarshalMsg(message)
+	if err != nil {
+		// if we received a message that we cannot parse, disconnect.
+		s.log.Infof("received unparsable transaction sync message from peer. disconnecting from peer.")
+		return err
+	}
+
+	if incomingMessage.message.Version != txnBlockMessageVersion {
```

Comment:
The support for older peers isn't an issue ( yet ). The version is the only placeholder for that, so we can safely interact with *newer* peers without any problem. This code is less likely to be invoked, since we would likely to use the Network Protocol versioning for that. However, it's good practice to have a low-level fallback.